addons/promote/common/src/main/java/org/commonjava/indy/promote/data/PromotionManager.java

```diff
@@ -761,430 +561,324 @@ public PathsPromoteResult rollbackPathsPromote( final PathsPromoteResult result
+            result.withPromotionId( result.getPromotionId() );
+
+            if ( result.getRequest().getCallback() != null )
+            {
+                return callbackHelper.callback( result.getRequest().getCallback(), result );
+            }
+
+            return result;
         } ) );
     }
 
-    private PathsPromoteResult doRollbackPathsPromote( PathsPromoteResult result )
-            throws IndyWorkflowException
+    private PathsPromoteResult doPathsPromotion( PathsPromoteRequest request, boolean skipValidation, String baseUrl )
+                    throws IndyWorkflowException, PromotionValidationException
     {
-        StoreKey targetKey = result.getRequest().getTarget();
-
-        final List<Transfer> contents = getTransfersForPaths( targetKey, result.getCompletedPaths() );
-        final Set<String> completed = result.getCompletedPaths();
-        final Set<String> skipped = result.getSkippedPaths();
+        final Set<String> paths = request.getPaths();
+        final StoreKey source = request.getSource();
 
-        if ( completed == null || completed.isEmpty() )
+        List<Transfer> contents;
+        if ( paths == null || paths.isEmpty() )
         {
-            result.setError( null );
-            return result;
+            contents = downloadManager.listRecursively( source, DownloadManager.ROOT_PATH );
         }
-
-        Set<String> pending =
-                        result.getPendingPaths() == null ? new HashSet<>() : new HashSet<>( result.getPendingPaths() );
-
-        final AtomicReference<String> error = new AtomicReference<>();
-        final boolean copyToSource = result.getRequest().isPurgeSource();
-
-        ArtifactStore source = null;
-        try
+        else
         {
-            source = storeManager.getArtifactStore( result.getRequest().getSource() );
+            contents = promotionHelper.getTransfersForPaths( source, paths );
         }
-        catch ( final IndyDataException e )
+
+        final Set<String> pending = new HashSet<>();
+        for ( final Transfer transfer : contents )
         {
-            String msg =
-                            String.format( "Failed to retrieve artifact store: %s. Reason: %s", result.getRequest().getSource(),
-                                           e.getMessage() );
+            pending.add( transfer.getPath() );
+        }
 
-            logger.error( msg, e );
-            error.set( msg );
+        if ( pending.isEmpty() )
+        {
+            return new PathsPromoteResult( request, pending, emptySet(), emptySet(), null );
         }
 
-        AtomicReference<IndyWorkflowException> wfEx = new AtomicReference<>();
-        if ( error.get() == null )
+        AtomicReference<Exception> ex = new AtomicReference<>();
+        StoreKeyPaths plk = new StoreKeyPaths( request.getTargetKey(), pending );
+
+        PathsPromoteResult promoteResult =
+                        byPathTargetLocks.lockAnd( plk, config.getLockTimeoutSeconds(), k ->
```

Comment:
I'm a little unclear on what this lock is actually helping us prevent now. It seems like the conflict manager is actually validating that the specific target locks are clear, and holding them until the promotion is done before unlocking them. The StoreKeyPaths instance would have to match exactly for this lock to mean anything.

Reply:
Well, I would agree on this byPathTargetLocks does not mean much here now. I will remove it to make the code a bit simple.